File: src/WindowStream.Core/Transport/PacketHeader.cs
using System;
using System.Buffers.Binary;

namespace WindowStream.Core.Transport;

public readonly record struct PacketHeader(
    uint StreamId,
    uint Sequence,
    ulong PresentationTimestampMicroseconds,
    PacketFlags Flags,
    byte FragmentIndex,
    byte FragmentTotal)
{
    public const int HeaderByteLength = 24;
    public const int MaximumPayloadByteLength = 1200;
    public const uint MagicValue = 0x57535452; // 'WSTR'

    public PacketHeader(
        int StreamId,
        int Sequence,
        long PresentationTimestampMicroseconds,
        PacketFlags Flags,
        int FragmentIndex,
        int FragmentTotal)
        : this(
            checked((uint)StreamId),
            checked((uint)Sequence),
            checked((ulong)PresentationTimestampMicroseconds),
            Flags,
            checked((byte)FragmentIndex),
            checked((byte)FragmentTotal))
    {
        if (FragmentTotal <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(FragmentTotal), FragmentTotal, "fragmentTotal must be at least 1");
        }
        if ((uint)FragmentIndex >= (uint)FragmentTotal)
        {
            throw new ArgumentOutOfRangeException(nameof(FragmentIndex), FragmentIndex, $"fragmentIndex must be less than fragmentTotal {FragmentTotal}");
        }
    }

    public bool IsIdrFrame => (Flags & PacketFlags.IdrFrame) != 0;
    public bool IsLastFragment => (Flags & PacketFlags.LastFragment) != 0;

    public void WriteTo(Span<byte> destination)
    {
        if (destination.Length < HeaderByteLength)
        {
            throw new ArgumentException(
                $"destination must be at least {HeaderByteLength} bytes, got {destination.Length}",
                nameof(destination));
        }
        BinaryPrimitives.WriteUInt32BigEndian(destination[0..4], MagicValue);
        BinaryPrimitives.WriteUInt32BigEndian(destination[4..8], StreamId);
        BinaryPrimitives.WriteUInt32BigEndian(destination[8..12], Sequence);
        BinaryPrimitives.WriteUInt64BigEndian(destination[12..20], PresentationTimestampMicroseconds);
        destination[20] = (byte)Flags;
        destination[21] = FragmentIndex;
        destination[22] = FragmentTotal;
        destination[23] = 0x00;
    }

    public static PacketHeader Parse(ReadOnlySpan<byte> source)
    {
        if (source.Length < HeaderByteLength)
        {
            throw new MalformedPacketException(
                $"packet is {source.Length} bytes, minimum is {HeaderByteLength}");
        }
        uint magic = BinaryPrimitives.ReadUInt32BigEndian(source[0..4]);
        if (magic != MagicValue)
        {
            throw new MalformedPacketException($"unexpected magic: 0x{magic:X8}");
        }
        uint streamId = BinaryPrimitives.ReadUInt32BigEndian(source[4..8]);
        uint sequence = BinaryPrimitives.ReadUInt32BigEndian(source[8..12]);
        ulong presentationTimestamp = BinaryPrimitives.ReadUInt64BigEndian(source[12..20]);
        byte flags = source[20];
        byte fragmentIndex = source[21];
        byte fragmentTotal = source[22];
        if (fragmentTotal == 0)
        {
            throw new MalformedPacketException("fragmentTotal must be at least 1");
        }
        if (fragmentIndex >= fragmentTotal)
        {
            throw new MalformedPacketException(
                $"fragmentIndex {fragmentIndex} is not less than fragmentTotal {fragmentTotal}");
        }
        return new PacketHeader(
            StreamId: streamId,
            Sequence: sequence,
            PresentationTimestampMicroseconds: presentationTimestamp,
            Flags: (PacketFlags)flags,
            FragmentIndex: fragmentIndex,
            FragmentTotal: fragmentTotal);
    }
}

File: viewer/WindowStreamViewer/app/src/main/kotlin/com/mtschoen/windowstream/viewer/transport/PacketHeader.kt
package com.mtschoen.windowstream.viewer.transport

class MalformedPacketException(message: String) : RuntimeException(message)

class PacketHeader internal constructor(
    val streamId: Int,
    val sequence: Long,
    val presentationTimestampMicroseconds: Long,
    val flags: Int,
    val fragmentIndex: Int,
    val fragmentTotal: Int,
    private val buffer: ByteArray,
    private val payloadOffset: Int,
    val payloadLength: Int
) {
    val isIdrFrame: Boolean get() = (flags and FLAG_IDR_FRAME) != 0
    val isLastFragment: Boolean get() = (flags and FLAG_LAST_FRAGMENT) != 0

    fun payloadAt(index: Int): Byte {
        require(index in 0 until payloadLength) { "payload index $index out of range 0..<$payloadLength" }
        return buffer[payloadOffset + index]
    }

    fun copyPayloadInto(destination: ByteArray, destinationOffset: Int) {
        System.arraycopy(buffer, payloadOffset, destination, destinationOffset, payloadLength)
    }

    companion object {
        const val HEADER_BYTE_LENGTH: Int = 24
        const val MAXIMUM_PAYLOAD_BYTE_LENGTH: Int = 1200
        const val MAGIC_VALUE: Int = 0x57535452
        const val FLAG_IDR_FRAME: Int = 0x01
        const val FLAG_LAST_FRAGMENT: Int = 0x02

        fun parse(buffer: ByteArray, length: Int): PacketHeader {
            if (length < HEADER_BYTE_LENGTH) {
                throw MalformedPacketException("packet of $length bytes shorter than header ($HEADER_BYTE_LENGTH)")
            }
            val magic: Int = readBigEndianInt(buffer, 0)
            if (magic != MAGIC_VALUE) {
                throw MalformedPacketException("unexpected magic: 0x${magic.toString(16)}")
            }
            val streamId: Int = readBigEndianInt(buffer, 4)
            val sequence: Long = readBigEndianInt(buffer, 8).toLong() and 0xFFFF_FFFFL
            val presentationTimestampMicroseconds: Long = readBigEndianLong(buffer, 12)
            val flags: Int = buffer[20].toInt() and 0xFF
            val fragmentIndex: Int = buffer[21].toInt() and 0xFF
            val fragmentTotal: Int = buffer[22].toInt() and 0xFF
            if (fragmentTotal == 0) {
                throw MalformedPacketException("fragmentTotal must be >= 1")
            }
            if (fragmentIndex >= fragmentTotal) {
                throw MalformedPacketException("fragmentIndex $fragmentIndex >= fragmentTotal $fragmentTotal")
            }
            val payloadLength: Int = length - HEADER_BYTE_LENGTH
            if (payloadLength > MAXIMUM_PAYLOAD_BYTE_LENGTH) {
                throw MalformedPacketException("payload $payloadLength exceeds maximum $MAXIMUM_PAYLOAD_BYTE_LENGTH")
            }
            return PacketHeader(
                streamId = streamId,
                sequence = sequence,
                presentationTimestampMicroseconds = presentationTimestampMicroseconds,
                flags = flags,
                fragmentIndex = fragmentIndex,
                fragmentTotal = fragmentTotal,
                buffer = buffer,
                payloadOffset = HEADER_BYTE_LENGTH,
                payloadLength = payloadLength
            )
        }

        private fun readBigEndianInt(buffer: ByteArray, offset: Int): Int {
            return ((buffer[offset].toInt() and 0xFF) shl 24) or
                ((buffer[offset + 1].toInt() and 0xFF) shl 16) or
                ((buffer[offset + 2].toInt() and 0xFF) shl 8) or
                (buffer[offset + 3].toInt() and 0xFF)
        }

        private fun readBigEndianLong(buffer: ByteArray, offset: Int): Long {
            var result = 0L
            for (index in 0 until 8) {
                result = (result shl 8) or (buffer[offset + index].toLong() and 0xFFL)
            }
            return result
        }
    }
}
